### PR TITLE
Fix optimized build

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
Fix building optimized build due to fPIC vs fPIE 

### Testing Done
Manual optimized builds

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Review by akhilveeraghanta — DISMISSED** [review]
I assume you've tested this and it improves build times, was just reading here https://stackoverflow.com/questions/3544035/what-is-the-difference-between-fpic-and-fpic-gcc-parameters and someone said it didn't really help in `x86`

File: src/cc_toolchain/cc_toolchain_config.bzl [hunk 1]
```diff
@@ -428,7 +428,7 @@ def _linux_gcc_impl(ctx):
                 flag_groups = [
                     flag_group(flags = ["-fPIC"], expand_if_available = "pic"),
                     flag_group(
-                        flags = ["-fPIE"],
+                        flags = ["-fPIC"],
                         expand_if_not_available = "pic",
                     ),
```
Comment (garethellis0) [inline src/cc_toolchain/cc_toolchain_config.bzl]: is this duplicating the above line? Why was this change needed?

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Sorry, just a bit confused about what

**Comment by MathewMacDougall** [comment]
>I assume you've tested this and it improves build times

@akhilveeraghanta this is for building code with optimizations, which will slow the build time but give faster code.

I also realized this was/is breaking any optimized build, not just with coverage. I confirmed this change works yesterday during some testing.

**Comment by garethellis0** [comment]
@MathewMacDougall - just to be clear, after this change,we have:
```
                flag_groups = [
                    flag_group(
                            flags = ["-fPIC"], 
                            expand_if_available = "pic"
                            ),
                    flag_group(
                            flags = ["-fPIC"],
                            expand_if_not_available = "pic",
                            ),
                ],
```
Seems like we should just be deleting the `flag_group` instead?

**Comment by jonathanlew** [comment]
> @MathewMacDougall - just to be clear, after this change,we have:
> 
> ```
>                 flag_groups = [
>                     flag_group(
>                             flags = ["-fPIC"], 
>                             expand_if_available = "pic"
>                             ),
>                     flag_group(
>                             flags = ["-fPIC"],
>                             expand_if_not_available = "pic",
>                             ),
>                 ],
> ```
> 
> Seems like we should just be deleting the `flag_group` instead?

Cleaned it up

File: src/cc_toolchain/cc_toolchain_config.bzl (hunk 1 above)
Comment (jonathanlew) [reply src/cc_toolchain/cc_toolchain_config.bzl]: Not that it's not the same group: expand_if_not_available vs expand_if_available, but yes I can clean it up

Comment (garethellis0) [reply src/cc_toolchain/cc_toolchain_config.bzl]: Ah, I see the difference, but does this mean that if "pic" is not available, we use the `-fPIC` flag anyways?

File: src/cc_toolchain/cc_toolchain_config.bzl [hunk 2]
```diff
@@ -426,11 +426,7 @@ def _linux_gcc_impl(ctx):
             flag_set(
                 actions = [ACTION_NAMES.c_compile, ACTION_NAMES.cpp_compile],
                 flag_groups = [
-                    flag_group(flags = ["-fPIC"], expand_if_available = "pic"),
-                    flag_group(
-                        flags = ["-fPIE"],
-                        expand_if_not_available = "pic",
-                    ),
+                    flag_group(flags = ["-fPIC"]),
```
Comment (MathewMacDougall) [inline src/cc_toolchain/cc_toolchain_config.bzl:429]: Just double checking, now we don't need `expand_if_available` at all?

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
LGTM

File: src/cc_toolchain/cc_toolchain_config.bzl (hunk 2 above)
Comment (jonathanlew) [reply src/cc_toolchain/cc_toolchain_config.bzl:429]: yeah since we do the same thing (fPIC) whether pic is available or not

File: src/cc_toolchain/cc_toolchain_config.bzl (hunk 1 above)
Comment (jonathanlew) [reply src/cc_toolchain/cc_toolchain_config.bzl]: yes I think so

Comment (garethellis0) [reply src/cc_toolchain/cc_toolchain_config.bzl]: Sorry, I'm still a bit confused here then. Why are we using position independent code when that's not available?

Comment (jonathanlew) [reply src/cc_toolchain/cc_toolchain_config.bzl]: "expand_if_not_available: A build variable that needs to be unavailable in order for this flag_group to be expanded."
So in the past we used `fPIE` if the `pic` build variable was not available, but that I don't think that `pic` being unavailable means we can't compile it with `fPIC`. What are your thoughts?

Comment (garethellis0) [reply src/cc_toolchain/cc_toolchain_config.bzl]: How are these build variables set? What does "pic" not being set mean? I'm guessing that either we should fix pic not being set when it should be, or just not use any flag if "pic" is not available, but really all depends on the answer to my first question here.

Comment (jonathanlew) [reply src/cc_toolchain/cc_toolchain_config.bzl]: iinm pic is set when it is compiled with our toolchain since it will have `pic_feature` with `name = "pic"`, so the only situations where we have `pic` unavailable is from 3rd party libraries, which should have pic since they should be libraries. I'm not sure if that makes sense and if we can somehow test my hypothesis

**Comment by jonathanlew** [comment]
@MathewMacDougall @garethellis0 Any further changes required?

**Comment by MathewMacDougall** [comment]
LGTM but @garethellis0 is more knowledgable about the toolchain than me. Just double checking, we don't need any of the `expand_if_not_available` lines anymore? It's fine to just have the new flag? Seems like we removed "fallback behavior"

**Review by garethellis0 — APPROVED** [review]
My apologies for the delay on this, dropped off my radar.

From `man gcc`:
```
If supported for the target machine, emit position-independent code, suitable for dynamic linking and avoiding any limit on the size of the global offset table.  This
           option makes a difference on AArch64, m68k, PowerPC and SPARC
```

So looks like `gcc` will do the leg work of checking if this is actually supported before using it. So I'm happy to merge.

LGTM :+1: :tada:

**Review by MathewMacDougall — APPROVED** [review]
